### PR TITLE
controller: Don't allow artifact type or uri to be empty strings and fix validation error responses

### PR DESCRIPTION
**Pull request by jvatic**

### CODE REVIEW

File: controller/controller.go [hunk 1]
```diff
@@ -80,8 +81,16 @@ type handlerConfig struct {
 
 // NOTE: this is temporary until httphelper supports custom errors
 func respondWithError(w http.ResponseWriter, err error) {
-	switch err.(type) {
+	switch v := err.(type) {
 	case ct.ValidationError:
+		detail := map[string]string{"field": v.Field}
+		raw, _ := json.Marshal(detail)
+		rawMessage := json.RawMessage(raw)
```
Comment (titanous) [inline controller/controller.go]: This step should be unnecessary.

File: controller/controller.go [hunk 2]
```diff
@@ -80,8 +81,16 @@ type handlerConfig struct {
 
 // NOTE: this is temporary until httphelper supports custom errors
 func respondWithError(w http.ResponseWriter, err error) {
-	switch err.(type) {
+	switch v := err.(type) {
 	case ct.ValidationError:
+		detail := map[string]string{"field": v.Field}
+		raw, _ := json.Marshal(detail)
```
Comment (titanous) [inline controller/controller.go]: Pass `detail` inline here.

**Comment by jvatic** [comment]
@titanous comments addressed

File: controller/controller_test.go [hunk 3]
```diff
@@ -179,6 +180,12 @@ func (s *S) TestProtectedApp(c *C) {
 }
 
 func (s *S) createTestArtifact(c *C, in *ct.Artifact) *ct.Artifact {
+	if in.Type == "" {
+		in.Type = "docker"
+	}
+	if in.URI == "" {
+		in.URI = fmt.Sprintf("https://example.com/%d", time.Now().Unix())
```
Comment (titanous) [inline controller/controller_test.go]: Use a random string generated by `pkg/random`

**Comment by titanous** [comment]
LGTM

**Pull request merged**